### PR TITLE
#23 - Add source comments from file descriptor

### PR DESCRIPTION
**Pull request by n3rdyme**
Added comments from fileDescriptor.sourceCodeInfo to enums, message interfaces, and services.

Should resolve issue 23
https://github.com/stephenh/ts-proto/issues/23

### CODE REVIEW

File: src/sourceInfo.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    },
+    enum: {
+        value: 2,
+    },
+    service: {
+        method: 2,
+    },
+}
+
+export interface SourceDescription {
```
Comment (stephenh) [inline src/sourceInfo.ts]: Can you add a link to the docs (assuming they exist, maybe the protobuf declaration file?) of what these fields are?

File: src/sourceInfo.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
+import { google } from '../build/pbjs';
+import FileDescriptorProto = google.protobuf.FileDescriptorProto;
+
+export type FieldID = number;
+
+export const Fields = {
+    file: {
```
Comment (stephenh) [inline src/sourceInfo.ts]: I don't have it setup to auto-run, but there is a `.prettierrc` checked in; can you run prettier on this?

File: src/sourceInfo.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+class EmptyDescription implements SourceDescription {
+    span = [];
+    leadingComments = '';
+    trailingComments = '';
+    leadingDetachedComments = [];
+};
+
+export type SourceInfoMap = {
+    [key: string]: SourceDescription
```
Comment (stephenh) [inline src/sourceInfo.ts]: Maybe add an in-source comment about what some example keys are?

File: src/sourceInfo.ts [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    get span() { return this.selfDescription.span; }
+    get leadingComments() { return this.selfDescription.leadingComments; }
+    get trailingComments() { return this.selfDescription.trailingComments; }
+    get leadingDetachedComments() { return this.selfDescription.leadingDetachedComments; }
+
+    lookup(type: FieldID, index: number): SourceDescription {
+        return this.sourceCode[`${type}.${index}`] || new EmptyDescription();
+    }
+
+    open(type: FieldID, index: number): SourceInfo {
```
Comment (stephenh) [inline src/sourceInfo.ts]: What's the different between `lookup` and `open`? It looks like open...finds multiple results? Add in-source method comments for each?

File: src/sourceInfo.ts (hunk 1 above)
Comment (n3rdyme) [reply src/sourceInfo.ts]: @stephenh Done.

File: src/sourceInfo.ts (hunk 2 above)
Comment (n3rdyme) [reply src/sourceInfo.ts]: @stephenh Done.

File: src/sourceInfo.ts (hunk 3 above)
Comment (n3rdyme) [reply src/sourceInfo.ts]: @stephenh Done.

File: src/sourceInfo.ts (hunk 4 above)
Comment (n3rdyme) [reply src/sourceInfo.ts]: @stephenh Added some comments... generally lookup is get one thing, open is open the type so I can lookup more relative to that type. Another way to say it, Message has nested types, so use `open`, for everything else you can just use `lookup`.

**Comment by n3rdyme** [comment]
@stephenh 
I added the requested comments in simple.proto and cleaned up the conditional logic a bit. I assume you did not want me to regenerate all the binary and source files on branch. It seems like that might make merging difficult, so I assume you are keeping PR branches clean?

Otherwise, I think I covered all the asks above. Let me know if you need any additional changes.

**Comment by stephenh** [comment]
Looks great, thanks! I do actually check in the generated source files; I know in downstream / real projects that is an anti-pattern but given ts-proto's job is producing that output, I basically treat it as an mini-integration test to see how the output has changed in the PRs/diffs/etc. I'll go ahead and do a follow up commit with those changes now.

**Pull request merged**